src/core/lib/event_engine/posix_engine/posix_engine.cc

```diff
@@ -496,143 +493,74 @@ EventEngine::TaskHandle PosixEventEngine::RunAfterInternal(
 #if GRPC_ARES == 1 && defined(GRPC_POSIX_SOCKET_TCP)
 
 PosixEventEngine::PosixDNSResolver::PosixDNSResolver(
-    const ResolverOptions& options, PosixEnginePollerManager* poller_manager,
+    const ResolverOptions& options, PosixEventPoller* poller,
     std::shared_ptr<EventEngine> event_engine)
     : options_(options),
       event_engine_(std::move(event_engine)),
-      poller_manager_(poller_manager) {}
-
-PosixEventEngine::PosixDNSResolver::~PosixDNSResolver() {
-  grpc_core::MutexLock lock(&mu_);
-  // The DNSResolver is held alive by its caller (e.g.
-  // event_engine_client_channel_resolver). The caller should be alive until
-  // its passed-in on_resolve closure gets destroyed (usually through a
-  // RefCountedPtr embedded in the closure). Until that happened, the
-  // DNSResolver is alive as well as the EventEngine (since DNSResolver holds a
-  // shared_ptr to the engine). Therefore, when the DNSResolver is destroyed,
-  // all on_resolves should have been destroyed (they may not be called when the
-  // request was cancelled) and all inflight handles should have been cleared
-  // regardless of the actual sequence of events.
-  GPR_ASSERT(GPR_LIKELY(inflight_requests_.empty()));
-}
+      poller_(poller) {}
 
-LookupTaskHandle PosixEventEngine::PosixDNSResolver::LookupHostname(
+void PosixEventEngine::PosixDNSResolver::LookupHostname(
     LookupHostnameCallback on_resolve, absl::string_view name,
-    absl::string_view default_port, Duration timeout) {
-  PosixEventPoller* poller = poller_manager_->Poller();
-  GPR_ASSERT(poller != nullptr);
-  absl::StatusOr<GrpcAresHostnameRequest*> request =
-      GrpcAresHostnameRequest::Create(
-          name, default_port, options_.dns_server, timeout,
-          std::make_unique<GrpcPolledFdFactoryPosix>(poller),
-          event_engine_.get());
-  if (!request.ok()) {
-    // Report back initialization failure through on_resolve.
-    event_engine_->Run(
-        [on_resolve = std::move(on_resolve),
-         status = request.status()]() mutable { on_resolve(status); });
-    return LookupTaskHandle::kInvalid;
-  }
-  LookupTaskHandle handle{reinterpret_cast<intptr_t>(*request),
-                          aba_token_.fetch_add(1)};
-  {
-    grpc_core::MutexLock lock(&mu_);
-    inflight_requests_.insert(handle);
+    absl::string_view default_port) {
+  if (!ares_resolver_) {
+    auto status_or = AresResolver::CreateAresResolver(
```

Comment:
Done. That makes sense. Creating the `AresResolver` in the ctor allows us to simplify code further a bit and remove a few member variables as well. The type of the `ares_resolver_` needs to be changed to `absl::StatusOr<grpc_core::OrphanablePtr<AresResolver>>` to store the error status when `CreateAresResolver` fails and to report the status back in each lookup method.

Reply:
I don't mind doing that. Probably in a separate PR since it would require changing a few places including `net/grpc` as well. It would make a faster propagation of the error status in the failure case of creating the DNSResolver and probably make the structure of this code a bit better.

We would need to change `EventEngineClientChannelDNSResolver::StartRequest()` to call `PollingResolver::OnRequestComplete()` with the error status set in the `Resolver::Result` and probably return a `nullptr`. I'm not sure yet if an inline call of `PollingResolver::StartRequest()` -> `PollingResolver::OnRequestComplete()` will cause any issue or not. If it would we might need to put `OnRequestComplete()` inside a `event_engine->Run()`.

@drfloob WDYT?

Reply:
@markdroth The original design (~2.5 years ago) returned `absl::StatusOr<DNSResolver>`, but I was argued out of that. I think the main argument is that it's not really a recoverable scenario, so returning a nullptr would be sufficient.

Reply:
Sent https://github.com/grpc/grpc/pull/33744 for this.

Reply:
I've approved https://github.com/grpc/grpc/pull/33744.  Shall we get that merged first and then merge those changes into this PR?